ext/nnstreamer/tensor_decoder/tensordec-protobuf.cc

```diff
@@ -0,0 +1,166 @@
+ * @brief       NNStreamer tensor-decoder subplugin, "protobuf",
+ *              which converts tensor or tensors to Protocol Buffers.
+ * @see         https://github.com/nnstreamer/nnstreamer
+ * @author      Yongjoo Ahn <yongjoo1.ahn@samsung.com>
+ * @bug         No known bugs except for NYI items
+ *
+ */
+
+#include <glib.h>
+#include <gst/gstinfo.h>
+#include <nnstreamer_plugin_api_decoder.h>
+#include <nnstreamer_plugin_api.h>
+
+#include "nnstreamer.pb.h"    /* Generated by `protoc` */
+
+void init_pb (void) __attribute__ ((constructor));
+void fini_pb (void) __attribute__ ((destructor));
+
+#define DECODER_PB_BIN_CAPS_STR "application/octet-stream"
+
+/**
+ * @brief tensordec-plugin's GstTensorDecoderDef callback
+ */
+static int
+pb_init (void **pdata)
+{
+  /**
+   * Verify that the version of the library we linked is
+   * compatibile with the headers.
+   */
+  GOOGLE_PROTOBUF_VERIFY_VERSION;
+  *pdata = NULL; /* no private data are needed for this sub-plugin */
+  return TRUE;
+}
+
+/** 
+ * @brief tensordec-plugin's GstTensorDecoderDef callback 
+ */
+static void
+pb_exit (void **pdata)
+{
+  google::protobuf::ShutdownProtobufLibrary ();
+  return;
+}
+
+/**
+ * @brief tensordec-plugin's GstTensorDecoderDef callback
+ */
+static int
+pb_setOption (void **pdata, int opNum, const char *param)
+{
+  return TRUE;
+}
+
+/**
+ * @brief tensordec-plugin's GstTensorDecoderDef callback
+ */
+static GstCaps *
+pb_getOutCaps (void **pdata, const GstTensorsConfig * config)
+{
+  return gst_caps_from_string (DECODER_PB_BIN_CAPS_STR);
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static GstFlowReturn
+pb_decode (void **pdata, const GstTensorsConfig * config,
+    const GstTensorMemory * input, GstBuffer * outbuf)
+{
+  GstMapInfo out_info;
+  GstMemory *out_mem;
+  size_t size;
+
+  NNStreamer::Tensors tensors;
+  NNStreamer::Tensors::frame_rate *fr;
+
+  const unsigned int num_tensors = config->info.num_tensors;
+  g_assert (num_tensors > 0);
+  tensors.set_num_tensor (num_tensors);
+
+  fr = tensors.mutable_fr ();
+  fr->set_rate_n (config->rate_n);
+  fr->set_rate_d (config->rate_d);
+  
+  for (unsigned int i = 0; i < num_tensors; ++i) {
+    NNStreamer::Tensor *tensor = tensors.add_tensor ();
+    gchar *name = NULL;
+
+    name = config->info.info[i].name;
+    if (name == NULL) {
+      tensor->set_name ("Anonymous");
+    } else {
+      tensor->set_name (name);
+    }
+
+    tensor->set_type ((NNStreamer::Tensor::Tensor_type)
+                          config->info.info[i].type);
+
+    for (int j = 0; j < NNS_TENSOR_RANK_LIMIT; ++j) {
+      tensor->add_dimension (config->info.info[i].dimension[j]);
+    }
+
+    tensor->set_data (input[i].data, (int) input[i].size);
+  }
+
+  g_assert (outbuf);
+
+  size = tensors.ByteSizeLong ();
+
+  if (gst_buffer_get_size (outbuf) == 0) {
+    out_mem = gst_allocator_alloc (NULL, size, NULL);
+  } else {
+    if (gst_buffer_get_size (outbuf) < size) {
```

Comment:
It seems that `gst_buffer_get_size` is repeatedly called. How about declaring its variable? For example
```
outbuf_size = gst_buffer_get_size (outbuf);
if (outbuf_size == 0) {
  ...
} else {
  if (outbuf_size < size)
    ...
}

```

Reply:
Thanks for the neat suggestion! I updated the commit.